Added global folder and multi-folder support

ReiMcCl opened the pull request:
Local folders can be configured using "%WORKSPACE%", configurable through constants folder.

Templates Path would now appear as below:

`"blueprint.templatesPath": [
    "templates", // Located in Program Files
    "%WORKSPACE%/blueprint-templates" // Located in Workspace Root
 ],`

reesemclean commented on package.json:
```diff
@@ -1,9 +1,8 @@
-    "Easily create files and folder of files from templates in Visual Studio Code",
-  "version": "1.5.0",
+  "description": "Easily create files and folder of files from templates in Visual Studio Code",
+  "version": "1.6.0",
```
Since this is a breaking change I'll make this the 2.0.0 release -- I can make that change during release though.

reesemclean commented on src/inputController.ts:
```diff
@@ -31,35 +31,40 @@ export class InputController {
+                try {
+                    templateNames = this.availableTemplateNames(templatePath);
+
+                    const templateObject: string[] = templateNames.map((str) => templatePath + "\\" + str);
```
Let's use path.join here to make sure we are OS agnostic (we already had to make that change for Windows support a little bit ago).

reesemclean commented on src/inputController.ts:
```diff
@@ -31,35 +31,40 @@ export class InputController {
+                    const templateObject: string[] = templateNames.map((str) => templatePath + "\\" + str);
+                    templates = templates.concat(templateObject);
+                } catch (error) {
+                    // return;
```
Just cleanup the comments here by removing them -- maybe log the error?

reesemclean commented on src/inputController.ts:
```diff
@@ -31,35 +31,40 @@ export class InputController {
-                reject(new Error(`${constants.ERROR_SETUP_MESSAGE_PREFIX} No templates found in: ${templateFolderPath}. Please see ${constants.README_URL} for information on setting up Blueprint in your project.`));
-                return;
+                reject(new Error(`${constants.ERROR_SETUP_MESSAGE_PREFIX} No templates found. Please see ${constants.README_URL} for information on setting up Blueprint in your project.`));
+                /// return;
```
Let's return here so any side-effects after the reject don't run (that's my understanding of promises at least).

reesemclean commented on src/extension.ts:
```diff
@@ -21,11 +21,22 @@ export function activate(context: vscode.ExtensionContext) {
+            .get("templatesPath") as string[];
+
+        const templateFolderPath: string[] = [];
+        for (const templatePath of templateFolderRelativePath){
```
Can we map over templateFolderRelativePath instead of a for loop?

reesemclean commented on src/fileCreator.ts:
```diff
@@ -107,7 +106,7 @@ export class FileCreator {
         return new Promise((resolve, reject) => {
 
-            const templateDirectory = path.join(this.data.templateFolderPath, this.data.templateName);
+            const templateDirectory = path.join(this.data.templateFolderPath);
```
We don't need the path.join here anymore do we?

reesemclean commented:
I actually think we can just use relative paths — to reference a file inside the workspace the user should be able to use "./blueprint-templates”. So we should be able to get rid of %WORKSPACE% completely.

> On Apr 30, 2018, at 8:24 PM, Ray McClain <notifications@github.com> wrote:
> 
> @ReiMcCl commented on this pull request.
> 
> In src/fileCreator.ts <https://github.com/reesemclean/blueprint/pull/42#discussion_r185160920>:
> 
> > @@ -107,7 +106,7 @@ export class FileCreator {
>  
>          return new Promise((resolve, reject) => {
>  
> -            const templateDirectory = path.join(this.data.templateFolderPath, this.data.templateName);
> +            const templateDirectory = path.join(this.data.templateFolderPath);
> Thanks for the quick response! I believe we can alter the above to make this a non-breaking change by altering the behaviour to use a character/key for referencing the global directory instead of the local one. Something like "@/templates" referencing the program files. This way, we should be able to retain the existing functionality.
> 
> I will submit a new PR shortly with the above changes and the reversed logic for global/local directory.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/reesemclean/blueprint/pull/42#discussion_r185160920>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAz4fcKvCJ0YlDMrfRqJ8Wmfxi8JKb-Gks5tt9VngaJpZM4TsMa0>.
>

ReiMcCl commented:
I have added relative loading through "./", unfortunately the method still uses a key defined in constants as the relative path is from VSCode's program files.

I have also updated the templatesPath setting to support String, Object, and Array types. Backwards compatibility is an option, however it would require some type of global key like '@/'.
